### PR TITLE
Update to 1.82 and 1.83

### PR DESCRIPTION
**Pull request by Veykril**
Builds atop of #529

### CODE REVIEW

File: src/types-and-traits.rst [hunk 1]
```diff
@@ -2923,7 +2990,7 @@ Lifetimes
 .. syntax::
 
    Lifetime ::=
-       $$'$$ NonKeywordIdentifier
+       $$'$$  (NonKeywordIdentifier | RawIdentifierKeyword)
```
Comment (tshepang) [inline src/types-and-traits.rst:2993]:
```suggestion
       $$'$$ (NonKeywordIdentifier | RawIdentifierKeyword)
```

File: src/types-and-traits.rst [hunk 2]
```diff
@@ -2177,6 +2204,46 @@ be modified through :t:`[immutable reference]s`.
...
+  The :t:`type` is a :t:`struct type` with at least one :t:`visible <visibility>` :t:`field` whose :t:`type` is subject to :t:`visible emptiness`.
+
+* :dp:`fls_R65LmGTvSlIm`
+  The :t:`type` is a :t:`tuple type` with at least one of the contained :t:`[type]s` is subject to :t:`visible emptiness`.
+
+* :dp:`fls_yiPv3wKYkSRl`
+  The :t:`type` is an :t:`array type` with a non-zero :t:`size operand` and an :t:`element type` that is subject to :t:`visible emptiness`.
+
+:dp:`fls_S9QL6yVF5LFI`
+A :t:`enum variant` is subject to :t:`visible emptiness` when the :t:`type` of at least one of the :t:`enum variant`'s visible <visibility> :t:`[field]s` is subject to :t:`visible emptiness`.
```
Comment (tshepang) [inline src/types-and-traits.rst:2245]:
```suggestion
A :t:`enum variant` is subject to :t:`visible emptiness` when the :t:`type` of at least one of the :t:`enum variant`'s :t:`visible <visibility>` :t:`[field]s` is subject to :t:`visible emptiness`.
```

File: src/types-and-traits.rst [hunk 3]
```diff
@@ -2177,6 +2204,46 @@ be modified through :t:`[immutable reference]s`.
...
+A :t:`type` is subject to :t:`visible emptiness` as follows:
+
+* :dp:`fls_ZfnYORORz40y`
+  The :t:`type` is the :t:`never type`.
+
+* :dp:`fls_TE2GBnZX2YFW`
+  The :t:`type` is a :t:`zero-variant enum type`.
+
+* :dp:`fls_GNFzLMH8RBUC`
+  The :t:`type` is an :t:`enum type` not subject to :t:`attribute` :c:`non_exhaustive` with all :t:`[enum variant]s` are subject to :t:`visible emptiness`.
```
Comment (tshepang) [inline src/types-and-traits.rst:2233]: is this what is meant
```suggestion
  The :t:`type` is an :t:`enum type` not subject to :t:`attribute` :c:`non_exhaustive` where all :t:`[enum variant]s` are subject to :t:`visible emptiness`.
```

File: src/macros.rst [hunk 4]
```diff
@@ -840,7 +841,7 @@ A :t:`metavariable` in a :t:`macro matcher` is matched against a sequence of
   :t:`Fragment specifier` **block** requires a :t:`block expression`.
 
 * :dp:`fls_k00bck2k8tde`
-  :t:`Fragment specifier` **expr** requires an :t:`expression`.
+  :t:`Fragment specifier` **expr** and **expr** requires an :t:`expression` that is not an :t:`underscore expression` or :t:`const block expression`.
```
Comment (tshepang) [inline src/macros.rst:844]:
```suggestion
  :t:`Fragment specifier` **expr** and **expr_2021** requires an :t:`expression` that is not an :t:`underscore expression` or :t:`const block expression`.
```

File: src/patterns.rst [hunk 5]
```diff
@@ -90,12 +90,11 @@ Refutability
...
 
 :dp:`fls_9fjspnefoyvz`
-An :t:`irrefutable pattern` is a :t:`pattern` that always matches the :t:`value`
-it is being matched against.
+An :t:`irrefutable pattern` is a :t:`pattern` that always matches any :t:`value` of its :t:`type`.
 
 :dp:`fls_uq7ftuuq1sig`
 A :t:`refutable pattern` is a :t:`pattern` that has a possibility of not
-matching the :t:`value` it is being matched against.
+matching a :t:`value` of a its :t:`type`
```
Comment (tshepang) [inline src/patterns.rst:97]:
```suggestion
matching a :t:`value` of its :t:`type`
```

File: src/patterns.rst [hunk 6]
```diff
@@ -385,12 +384,10 @@ to an :t:`associated constant`.
...
 
 * :dp:`fls_cfoy86mkmqa4`
-  A :t:`constant` whose :t:`type` is an :t:`irrefutable pattern`, or
+  An :t:`irrefutable constant`,
 
 * :dp:`fls_rnppz6y5z8pi`
-  An :t:`enum variant` of an :t:`enum type` with a single :t:`enum variant`
-  where the :t:`[type]s` of all :t:`[field]s` are :t:`[irrefutable pattern]s`,
-  or
+  An :t:`enum variant` of an :t:`enum type` with zero or more :t:`[visibly empty enum variant]s` and one non-:t:`visibly empty enum variant` where the :t:`[type]s` of all :t:`[field]s` are not subject to :t:`visible emptiness`, or
```
Comment (tshepang) [inline src/patterns.rst:390]:
```suggestion
  An :t:`enum variant` of an :t:`enum type` with zero or more :t:`[visible empty enum variant]s` and one non-:t:`visible empty enum variant` where the :t:`[type]s` of all :t:`[field]s` are not subject to :t:`visible emptiness`
```

File: src/patterns.rst [hunk 7]
```diff
@@ -756,7 +753,7 @@ A :t:`struct pattern` is an :t:`irrefutable pattern` if
     A :t:`struct type` or a :t:`union type`, or
 
   * :dp:`fls_MK83WE0iDqNf`
-    An :t:`enum variant` of an :t:`enum` with a single :t:`enum variant`.
+    An :t:`enum variant` of an :t:`enum type` with zero or more :t:`[visibly empty enum variant]s` and one non-:t:`visibly empty enum variant` where the :t:`[type]s` of all :t:`[field]s` are not subject to :t:`visible emptiness`.
```
Comment (tshepang) [inline src/patterns.rst:756]:
```suggestion
    An :t:`enum variant` of an :t:`enum type` with zero or more :t:`[visible empty enum variant]s` and one non-:t:`visible empty enum variant` where the :t:`[type]s` of all :t:`[field]s` are not subject to :t:`visible emptiness`.
```

**Review by tshepang — CHANGES_REQUESTED** [review]

File: src/types-and-traits.rst (hunk 3 above)
Comment (Veykril) [reply src/types-and-traits.rst:2233]: yes

**Review by tshepang — APPROVED** [review]
bors merge

**Pull request merged**